### ISSUE TITLE
Unable to log-in with single sign-on enabled on CLM 6.0.4

### ISSUE THREAD

**Issue by edinev** [dixudx/rtcclient#112]
With a recent update to CLM 6.0.4 a single sign-on feature was enabled for our RTC servers. Before the update everything was working correctly using this client, but now I am unable to log in.
Here is a log:

2018-05-21 17:42:19,767 DEBUG client.RTCClient: Get response from https://host:9443/ccm3/authenticated/identity
2018-05-21 17:42:19,829 DEBUG urllib3.connectionpool: Starting new HTTPS connection (1): host
2018-05-21 17:42:20,204 DEBUG urllib3.connectionpool: https://host:9443 "GET /ccm3/authenticated/identity HTTP/1.1" 302 0
2018-05-21 17:42:20,219 ERROR client.RTCClient: Failed GET request at <https://host:9443/ccm3/authenticated/identity> with response: b''
2018-05-21 17:42:20,219 DEBUG client.RTCClient: Post a request to https://host:9443/ccm3/authenticated/j_security_check with data: j_username=user&j_password=pass and json: None
2018-05-21 17:42:20,219 DEBUG urllib3.connectionpool: Starting new HTTPS connection (1): host
2018-05-21 17:42:20,407 DEBUG urllib3.connectionpool: https://host:9443 "POST /ccm3/authenticated/j_security_check HTTP/1.1" 302 0
2018-05-21 17:42:20,407 ERROR client.RTCClient: Failed POST request at <https://host:9443/ccm3/authenticated/j_security_check> with response: b''
2018-05-21 17:42:20,422 INFO client.RTCClient: 302
2018-05-21 17:42:20,422 DEBUG client.RTCClient: Get response from https://host:9443/ccm3/authenticated/identity
2018-05-21 17:42:20,422 DEBUG urllib3.connectionpool: Starting new HTTPS connection (1): host
2018-05-21 17:42:20,516 DEBUG urllib3.connectionpool: https://host:9443 "GET /ccm3/authenticated/identity HTTP/1.1" 302 0
2018-05-21 17:42:20,563 ERROR client.RTCClient: Failed GET request at <https://host:9443/ccm3/authenticated/identity> with response: b''
Traceback (most recent call last):
  File "C:\Users\user\Desktop\SCRUM\script - Copy.py", line 53, in <module>
    myClient = RTCClient(url, username, password, ends_with_jazz = False)
  File "C:\Users\user\AppData\Local\Programs\Python\Python36\lib\site-packages\rtcclient\client.py", line 65, in __init__
    self.headers = self._get_headers()
  File "C:\Users\user\AppData\Local\Programs\Python\Python36\lib\site-packages\rtcclient\client.py", line 125, in _get_headers
    _headers["Cookie"] += "; " + resp.headers.get("set-cookie")
TypeError: must be str, not NoneType

It seems that there is something wrong with the authentication urls and the script isn't finding certain attributes within those urls.

**Comment by dixudx** [comment]
>  2018-05-21 17:42:20,219 ERROR client.RTCClient: Failed GET request at https://host:9443/ccm3/authenticated/identity with response: b''

Truncated log here?

> File "C:\Users\user\AppData\Local\Programs\Python\Python36\lib\site-packages\rtcclient\client.py", line 125, in _get_headers
_headers["Cookie"] += "; " + resp.headers.get("set-cookie")
TypeError: must be str, not NoneType

Seems the response header does not have field `set-cookie`.

@edinev Actually I've never tested this library against SSO feature. Would you please send me a full log? Did not find anything useful from truncated logs.

**Comment by edinev** [comment]
@dixudx Unfortunately the logs aren't truncated there just isn't a normal response from the get requests in question.
After some tinkering with it, I found out that when you try to access RTC server the link redirects to this url: https://host:9643/jazzop/form/login . After a successful login you are redirected back to the original RTC server.
I also noticed that there is a similar to the used in the library form at https://host:9643/jazzop/j_security_check where you post your security credentials.

**Comment by kmilev** [comment]
I have the same issue, is there any development ?

**Comment by dixudx** [comment]
@edinev @kmilev I did not get any detailed log on this. 

> I found out that when you try to access RTC server the link redirects.

https://github.com/dixudx/rtcclient/blob/1721dd0b047478f5bdd6359b07a2c503cfafd86f/rtcclient/client.py#L80-L83

What if you set `ends_with_jazz=True` to allow redirects.

> I also noticed that there is a similar to the used in the library form at https://host:9643/jazzop/j_security_check where you post your security credentials.

Any differences on the auth dict?

**Comment by kmilev** [comment]
@dixudx This is the log i get and it is basically the same (with the ends_with_jazz=True):

2018-05-29 12:55:07,473 DEBUG client.RTCClient: Get response from https://host:9443/ccm3/authenticated/identity
2018-05-29 12:55:07,493 DEBUG requests.packages.urllib3.connectionpool: Starting
 new HTTPS connection (1): host
2018-05-29 12:55:07,983 DEBUG requests.packages.urllib3.connectionpool: https://
host:9443 "GET /ccm3/authenticated/identity HTTP/1.1" 302 0
2018-05-29 12:55:08,063 DEBUG requests.packages.urllib3.connectionpool: https://
host:9443 "GET /ccm3/authenticated_identity HTTP/1.1" 401 0
2018-05-29 12:55:08,073 ERROR client.RTCClient: Failed GET request at <https://host
:9443/ccm3/authenticated/identity> with response:
Traceback (most recent call last):
  File "TryLogin.py", line 14, in <module>
    myclient = RTCClient(url, username, password, ends_with_jazz=True)
  File "C:\Python27\lib\site-packages\rtcclient\client.py", line 65, in __init__

    self.headers = self._get_headers()
  File "C:\Python27\lib\site-packages\rtcclient\client.py", line 90, in _get_hea
ders
    allow_redirects=_allow_redirects)
  File "C:\Python27\lib\site-packages\rtcclient\utils.py", line 22, in wrapper
    return func(*args, **kwargs)
  File "C:\Python27\lib\site-packages\rtcclient\base.py", line 83, in get
    response.raise_for_status()
  File "C:\Python27\lib\site-packages\requests-2.12.4-py2.7.egg\requests\models.
py", line 893, in raise_for_status
    raise HTTPError(http_error_msg, response=self)
requests.exceptions.HTTPError: 401 Client Error: Unauthorized for url: https://host
:9443/ccm3/authenticated_identity

**Comment by dixudx** [comment]
> 2018-05-29 12:55:07,983 DEBUG requests.packages.urllib3.connectionpool: https://
host:9443 "GET /ccm3/authenticated/identity HTTP/1.1" 302 0
2018-05-29 12:55:08,063 DEBUG requests.packages.urllib3.connectionpool: https://
host:9443 "GET /ccm3/authenticated_identity HTTP/1.1" 401 0

The response code 401 showed you're not authorized. Please help check your login account.

**Comment by kmilev** [comment]
It is working when i log in manually.

**Comment by franz-95** [comment]
I encountered a similar problem. For me the same logging output was displayed. 

For me the fix was to add the auth=(self.username, self.password) tuple to the arguments of the self.get() function in line 86 and 117 of the client.py. I also removed the lines 96 through 104 as they were not needed. With that the authentication proceeded properly and I can now use the library as needed.

**Comment by dixudx** [comment]
> For me the fix was to ...

@franz-95 Would you please submit a PR to address this?

**Comment by franz-95** [comment]
I only have my RTC available so I don't know if my changes affect other users. I could work with a try/except to catch Error 401 but I don't think that is a nice way of doing it.

**Comment by Josespaul** [comment]
This issue is still not fixed. Log-in is still failing.

**Comment by babuenir** [comment]
> I encountered a similar problem. For me the same logging output was displayed.
> 
> For me the fix was to add the auth=(self.username, self.password) tuple to the arguments of the self.get() function in line 86 and 117 of the client.py. I also removed the lines 96 through 104 as they were not needed. With that the authentication proceeded properly and I can now use the library as needed.

Thanks for the quick fix. I needed "_allow_redirects = True" too. I could access jazz RTC with this now.

**Comment by dixudx** [comment]
@babuenir Thanks for your verification.

**Comment by gmalmeida** [comment]
@franz-95 

I have added the recommended tuple  as described:

> I encountered a similar problem. For me the same logging output was displayed.
> 
> For me the fix was to add the auth=(self.username, self.password) tuple to the arguments of the self.get() function in line 86 and 117 of the client.py. I also removed the lines 96 through 104 as they were not needed. With that the authentication proceeded properly and I can now use the library as needed.

```python3
resp = self.get(self.url + "/authenticated/identity",
                        verify=False,
                        headers=_headers,
                        proxies=self.proxies,
                        auth=(self.username, self.password),
                        allow_redirects=_allow_redirects)
```
However, I am still observing the following error:
```
2020-05-06 14:48:58,580 ERROR client.RTCClient: Failed GET request at <https://alm.harman.com/ccm/authenticated/identity> with response: b'
``` 

Our Jazz Server is using:

```
Jazz Foundation - Core Libraries | 6.0.6 | RJF-SERVER-I20180517-1708
JTS Patch is active. Contact your system administrator for details on fixes included in this patch in readme.txt. | 6.0.6 | iFix015
``` 
Did you ever have the chance to test this fix on **6.0.6**?

Thank you guys!

**Comment by dixudx** [comment]
> 2020-05-06 14:48:58,580 ERROR client.RTCClient: Failed GET request at <https://alm.harman.com/ccm/authenticated/identity> with response: b'

@gmalmeida More error context for this? Needs more lines of logs.

**Comment by gmalmeida** [comment]
```

 self.jazz =  True
 self.url =  https://alm.harman.com/ccm
 self.CONTENT_XML =  text/xml
 allow_redirects =  True

resp.headers =  {'X-Powered-By': 'Servlet/3.0', 'Content-Language': 'en-US', 'Keep-Alive': 'timeout=120, max=99', 'Connection': 'keep-alive', 'RVBD-CSH': '172.16.35.51', 'X-RBT-CLI': 'Name=HIKARB10; Ver=9.9.2;', 'X-OneAgent-JS-Injection': 'true', 'Set-Cookie': 'JSA_SESSION_IDENTITY=061c70ba-536f-42b6-9d74-4d01c6d62b29; Path=/ccm; Secure; HttpOnly, JSESSIONID=0000fGe5MFB6lM64IHlDLTKVqPJ:a8e8c295-d2d7-4082-a3dc-e8208c111b68; Path=/; Secure; HttpOnly', 'RVBD-SSH': '10.70.0.10', 'Strict-Transport-Security': 'max-age=31536000', 'Cache-Control': 'no-cache="set-cookie, set-cookie2"', 'Content-Length': '54', 'Content-Type': 'text/json; charset=utf-8', 'Expires': 'Thu, 01 Dec 1994 16:00:00 GMT', 'Date': 'Wed, 06 May 2020 12:48:57 GMT'}
2020-05-06 14:48:57,644 DEBUG client.RTCClient: Post a request to https://alm.harman.com/ccm/j_security_check with data: j_username=gabriel-user&j_password=gabriel-pass and json: None
2020-05-06 14:48:57,646 DEBUG urllib3.connectionpool: Starting new HTTPS connection (1): alm.harman.com:443
2020-05-06 14:48:57,964 DEBUG urllib3.connectionpool: https://alm.harman.com:443 "POST /ccm/j_security_check HTTP/1.1" 302 0
2020-05-06 14:48:58,027 DEBUG urllib3.connectionpool: https://alm.harman.com:443 "GET /ccm/ HTTP/1.1" 302 0
2020-05-06 14:48:58,088 DEBUG urllib3.connectionpool: https://alm.harman.com:443 "GET /ccm/web HTTP/1.1" 200 1062

resp.headers2 =  {'X-Powered-By': 'Servlet/3.0', 'Content-Language': 'en-US', 'Keep-Alive': 'timeout=120, max=98', 'Connection': 'keep-alive', 'X-XSS-Protection': '1; mode=block', 'X-RBT-CLI': 'Name=HIKARB10; Ver=9.9.2;', 'RVBD-CSH': '172.16.35.51', 'X-OneAgent-JS-Injection': 'true', 'Content-Encoding': 'gzip', 'Set-Cookie': 'JSA_AUTH_COMPLETE=""; Expires=Thu, 01 Dec 1994 16:00:00 GMT; Path=/; Secure, JSA_AUTH_COMPLETE=""; Expires=Thu, 01 Dec 1994 16:00:00 GMT; Path=/; Secure, JSA_SESSION_IDENTITY=5ab26b0f-8a38-4e9e-a5e8-d0943b583fcc; Path=/ccm; Secure; HttpOnly', 'RVBD-SSH': '10.70.0.10', 'Strict-Transport-Security': 'max-age=31536000', 'Cache-Control': 'no-cache="set-cookie, set-cookie2"', 'Content-Length': '1062', 'Content-Type': 'text/html; charset=UTF-8', 'Expires': 'Thu, 01 Dec 1994 16:00:00 GMT', 'Date': 'Wed, 06 May 2020 12:48:57 GMT'}
Authfailed =  None
```

**Comment by dixudx** [comment]
> 2020-05-06 14:48:57,964 DEBUG urllib3.connectionpool: https://alm.harman.com:443 "POST /ccm/j_security_check HTTP/1.1" 302 0

From the status code 302, it seems the url has been changed to a newer one. @gmalmeida 

@gmalmeida Please follow solution from @franz-95 , and have another try.

Quoted,
> For me the fix was to add the auth=(self.username, self.password) tuple to the arguments of the self.get() function in line 86 and 117 of the client.py. I also removed the lines 96 through 104 as they were not needed. With that the authentication proceeded properly and I can now use the library as needed.

https://github.com/dixudx/rtcclient/blob/master/rtcclient/client.py#L86-L117

Just modify the file `client.py` installed in site-packages folder. And have another try.

**Comment by dixudx** [comment]
@franz-95 Would you like to submit your PR? I would like to package a new version to fix this incompatible issue.

**Comment by gmalmeida** [comment]
@dixudx , that is exactly what I have done -> see my previous message.

Unfortunately, it did not work for me. Probably I am missing something else that I am not able to point out.


> @franz-95
> 
> I have added the recommended tuple as described:
> 
> > I encountered a similar problem. For me the same logging output was displayed.
> > For me the fix was to add the auth=(self.username, self.password) tuple to the arguments of the self.get() function in line 86 and 117 of the client.py. I also removed the lines 96 through 104 as they were not needed. With that the authentication proceeded properly and I can now use the library as needed.
> 
> ```python
> resp = self.get(self.url + "/authenticated/identity",
>                         verify=False,
>                         headers=_headers,
>                         proxies=self.proxies,
>                         auth=(self.username, self.password),
>                         allow_redirects=_allow_redirects)
> ```
> 
> However, I am still observing the following error:
> 
> ```
> 2020-05-06 14:48:58,580 ERROR client.RTCClient: Failed GET request at <https://alm.harman.com/ccm/authenticated/identity> with response: b'
> ```
> 
> Our Jazz Server is using:
> 
> ```
> Jazz Foundation - Core Libraries | 6.0.6 | RJF-SERVER-I20180517-1708
> JTS Patch is active. Contact your system administrator for details on fixes included in this patch in readme.txt. | 6.0.6 | iFix015
> ```
> 
> Did you ever have the chance to test this fix on **6.0.6**?
> 
> Thank you guys!

**Comment by dixudx** [comment]
@franz-95 @babuenir Would you like to elaborate the fix?

**Comment by babuenir** [comment]
@dixudx, @gmalmeida probably you would need to set `_allow_redirects = True` irrespective of the condition. In _get_headers method, I changed like below
```
        if self.jazz is True:
            _allow_redirects = True
        else:
            _allow_redirects = False
```
to
```
        _allow_redirects = True
```
In my usecase, the redirection to authentication page works by doing so.

**Comment by gmalmeida** [comment]
@babuenir, thanks for your inputs.
My [previous message](https://github.com/dixudx/rtcclient/issues/112#issuecomment-625052417)  suggests I have this properly set to `True`. So I think the error may be related to something else.

**Comment by MykolaKovalov** [comment]
Hi guys!
I have an issue same as @gmalmeida. 
Do someone have solution?
Thanks in advance!

**Comment by dixudx** [comment]
@franz-95 @babuenir  Would you like to submit a PR to address this issue? Thanks ahead.

I would like to package a newer version for this library. This fix is quite critical.

**Comment by k10blogger** [comment]
```
c:\Data\Workspace\HookScripts>py -2 rtc.py > debug_message.txt
2020-11-02 21:21:31,799 DEBUG client.RTCClient: Get response from https://urlhidden:9443/ccm/authenticated/identity
2020-11-02 21:21:31,806 DEBUG urllib3.connectionpool: Starting new HTTPS connect
ion (1): urlhidden:9443
2020-11-02 21:21:32,868 DEBUG urllib3.connectionpool: https://urlhidden:9443 "GET /ccm/authenticated/identity HTTP/1.1" 302 None
2020-11-02 21:21:33,078 DEBUG urllib3.connectionpool: https://urlhidden:9443 "GET /ccm/authenticated_identity HTTP/1.1" 401 0
2020-11-02 21:21:33,079 ERROR client.RTCClient: Failed GET request at <https://urlhidden:9443/ccm/authenticated/identity> with response:
Traceback (most recent call last):
  File "rtc.py", line 19, in <module>
    myclient = RTCClient(base_url,username,password)
  File "C:\Program Files\Python27\lib\site-packages\rtcclient\client.py", line 6
5, in __init__
    self.headers = self._get_headers()
  File "C:\Program Files\Python27\lib\site-packages\rtcclient\client.py", line 9
0, in _get_headers
    allow_redirects=_allow_redirects)
  File "C:\Program Files\Python27\lib\site-packages\rtcclient\utils.py", line 22
, in wrapper
    return func(*args, **kwargs)
  File "C:\Program Files\Python27\lib\site-packages\rtcclient\base.py", line 83,
 in get
    response.raise_for_status()
  File "C:\Program Files\Python27\lib\site-packages\requests\models.py", line 94
1, in raise_for_status
    raise HTTPError(http_error_msg, response=self)
requests.exceptions.HTTPError: 401 Client Error: Unauthorized for url: https://urlhidden:9443/ccm/authenticated_identity
```
I have redirection on and i am using 6.0.6.1. we have SSO but when it fails it redirects to a form-login. It seems there is some change in the way authentication is done in 6.0.6.1

**Comment by dixudx** [comment]
@k10blogger Please try the fix mentioned [above](https://github.com/dixudx/rtcclient/issues/112#issuecomment-525295452) (Credit to @franz-95  ).

**Comment by k10blogger** [comment]
Thanks that seems to work.

**Comment by oskrdt** [comment]
> I encountered a similar problem. For me the same logging output was displayed.
> 
> For me the fix was to add the auth=(self.username, self.password) tuple to the arguments of the self.get() function in line 86 and 117 of the client.py. I also removed the lines 96 through 104 as they were not needed. With that the authentication proceeded properly and I can now use the library as needed.

Is there any plan to include this fix in a PR?

**Comment by dixudx** [comment]
/close

Got fixed in #133.
